agent/spec/lib/kontena/observer_spec.rb

```diff
@@ -35,45 +55,92 @@ def crash
-      expect(subject).to_not be_ready
+    describe '#unwrap_observable' do
+      it 'returns the wrapped object' do
+        expect(described_class.unwrap_observable(observable)).to be_a TestObservable
+        expect(described_class.unwrap_observable(observable).class).to eq TestObservable
+      end
     end
 
-    it "immediately yields an updated value", :celluloid => true do
-      observable.update_observable object
+    describe '#observe' do
+      it "does not observe any value if not yet updated" do
+        subject.test_observe(observable)
 
-      subject
+        expect(subject).to_not be_ready
+      end
 
-      expect(subject).to be_ready
-      expect(subject.values).to eq [object]
-    end
+      it "immediately yields an updated value" do
+        observable.update_observable object
 
-    it "later yields after updating value", :celluloid => true do
-      subject
+        subject.test_observe(observable)
+
+        expect(subject).to be_ready
+        expect(subject.values).to eq [object]
+      end
 
-      expect(subject).to_not be_ready
+      it "later yields after updating value" do
+        subject.test_observe(observable)
 
-      observable.update_observable object
+        expect(subject).to_not be_ready
 
-      expect(subject).to be_ready
-      expect(subject.values).to eq [object]
+        observable.update_observable object
+
+        expect(subject).to be_ready
+        expect(subject.values).to eq [object]
+      end
+
+      it "crashes if the observable does", :log_celluloid_actor_crashes => false do
+        subject.test_observe(observable)
+
+        expect{observable.crash}.to raise_error(RuntimeError)
+
+        expect{subject.ready?}.to raise_error(Celluloid::DeadActorError)
+      end
     end
 
-    it "crashes if the observable does", :celluloid => true, :log_celluloid_actor_crashes => false do
-      expect{observable.crash}.to raise_error(RuntimeError)
+    describe '#observe_sync' do
+      it 'raises timeout if the observable is not ready' do
+        expect{
+          subject.observe(observable, timeout: 0.01)
+        }.to raise_error(Timeout::Error, 'timeout after waiting 0.01s until: Observable<TestObservable>')
+      end
+
+      it 'immediately returns value if updated' do
+        observable.update_observable(object)
+
+        expect(subject.observe(observable)).to eq object
+
+        observable.reset_observable
+      end
+
+      it 'blocks until observable' do
+        observable.delay_update(object, delay: 0.5)
 
-      expect{subject.ready?}.to raise_error(Celluloid::DeadActorError)
+        # NOTE: the class must include the WaitHelper, so that it uses Celluloid#sleep
+        #       if the wait_until! uses Kernel#sleep and blocks the actor thread,
+        #       then this spec will fail, because the delayed update doesn't have a chance to run
+        expect(subject.observe(observable, timeout: 1.0)).to eq object
+      end
+
+      it 'does not lose wait messages' do
```

Comment:
I'm pretty sure it would be impossible to make this spec pass using the `Celluloid#receive` API (at 29c4db32160b5ad53c2bd89343eec94275b64c51):

```
D, [2017-08-17T20:20:39.172119 #28489] DEBUG -- TestObservable: waiter: Observer<TestObserver>...
DEBUG: wait Observable<TestObservable>... (timeout=1.0)
D, [2017-08-17T20:20:39.272178 #28489] DEBUG -- TestObservable: update: 1
D, [2017-08-17T20:20:39.272511 #28489] DEBUG -- TestObservable: notify: Observer<TestObserver> <- 1...
D, [2017-08-17T20:20:39.282949 #28489] DEBUG -- TestObservable: update: 2
D, [2017-08-17T20:20:39.293702 #28489] DEBUG -- TestObservable: update: 3
...
D, [2017-08-17T20:20:39.662842 #28489] DEBUG -- TestObservable: update: 38
```

```
Failures:

  1) Kontena::Observer For a single observable #wait_observable does not lose wait messages
     Failure/Error: expect(subject.wait_observable!(observable, timeout: 1.0)).to be_a Integer
     
     Timeout::Error:
       timeout waiting 1.00s until: Observable<TestObservable> is ready
     # ./vendor/bundle/ruby/2.3.0/gems/celluloid-essentials-0.20.5/lib/celluloid/internals/responses.rb:29:in `value'
     # ./vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:45:in `value'
     # ./vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/proxy/sync.rb:22:in `method_missing'
     # ./spec/lib/kontena/observer_spec.rb:136:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:80:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:79:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:63:in `block (2 levels) in <top (required)>'
```

Reply:
This may not be the case anymore, now that the `observable.observe` is an atomic call protected by a mutex; once that returns, the observe task should be able to register the `Celluloid.receive` handler before any observable update messages arriving in the observing actor mailbox are processed.

Evidenced by the [`Celluloid::Future`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/future.rb#L90) also using `Celluloid.receive` in a similar fashion.

Reply:
This is what a race here looks like with celluloid logging, although it's more apparent in the mocked out `Kontena::Observer For two observables #observe_sync accepts update for first value while requesting second value` spec:

```
D, [2017-08-29T13:35:52.708039 #16208] DEBUG -- TestObserverActor: observe sync Kontena::Observable<TestObservableActor>...
D, [2017-08-29T13:35:52.708334 #16208] DEBUG -- Kontena::Observable<TestObservableActor>: update: #[Double :test1]
D, [2017-08-29T13:35:52.708381 #16208] DEBUG -- Kontena::Observable<TestObservableActor>: notify and drop: Kontena::Observer::Observe::Sync<TestObserverActor> <- #[Double :test1]
D, [2017-08-29T13:35:52.708454 #16208] DEBUG -- : Discarded message (unhandled): #<Kontena::Observable::Message:0x000000011c9ae8>
D, [2017-08-29T13:35:52.708555 #16208] DEBUG -- TestObserverActor: observe sync Kontena::Observable<TestObservableActor> => #[Double :test2]
D, [2017-08-29T13:35:52.708598 #16208] DEBUG -- TestObserverActor: observe wait Kontena::Observable<TestObservableActor>?, Kontena::Observable<TestObservableActor>... (wait timeout=)
```

Note that somehow celluloid processes the observable message before the call to receive, so it gets discarded... that leaves the observe stuck in the timeout.